File: WeeklyCode/week2/Remove Duplicates from Sorted List/chenhao.cpp
#include<iostream>
#include<algorithm>
#include<vector>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *deleteDuplicates(ListNode *head) {
        ListNode vHead(0);
        vHead.next=head;
        if(head!=NULL && head->val==0){
            vHead.val=1;
        }

        ListNode *pre=&vHead, *cur=head;
        while(cur!=NULL){
            ListNode* next=cur->next;
            if(cur->val == pre->val){
                pre->next=next;
                delete cur;
            }
            else{
                pre=cur;
            }
            cur=next;
        }
        return vHead.next;
    }
};

int main(){
    ListNode head(1);
    Solution s;
    cout<<s.deleteDuplicates(&head)<<endl;
}

File: WeeklyCode/week5/Convert Sorted List to Binary Search Tree/chenhao.java

public class Solution {
    
    private int maxDepth;
    private int nNodesInLastDepth;
    
    private ListNode pList;
    
    private TreeNode buildTree(int depth){
        if(depth>=maxDepth){
            if(nNodesInLastDepth>0){
                nNodesInLastDepth--;
                return new TreeNode(0);
            }
            else{
                return null;
            }
        }
        TreeNode node=new TreeNode(0);
        node.left=buildTree(depth+1);
        node.right=buildTree(depth+1);
        return node;
    }
    
    private void assignValues(TreeNode cur){
        if(cur==null || pList==null){
            return;
        }
        assignValues(cur.left);
        cur.val=pList.val;
        pList=pList.next;
        assignValues(cur.right);
    }
    
    public TreeNode sortedListToBST(ListNode head) {
        if(head==null){
            return null;
        }
        
        int len=0;
        ListNode cur=head;
        while(cur!=null){
            len++;
            cur=cur.next;
        }
        
        maxDepth=0;
        nNodesInLastDepth=len;
        while( nNodesInLastDepth > (1<<maxDepth) ){
            nNodesInLastDepth-=(1<<maxDepth);
            maxDepth++;
        }
        maxDepth++;
        
        TreeNode root=buildTree(1);
        
        pList=head;
        assignValues(root);
        
        return root;
    }
}